### PR TITLE
Add internal-only Python API for experimental matmul operations based on layout

### PR DESCRIPTION
**Pull request by kevinstephano**
The following operations are added to the Python API:

-  `ops._matmul_nn`
-  `ops._matmul_nt`
-  `ops._matmul_tn`
-  `ops._matmul_tt`

These operations have the following restrictions:

- They are dependent on sizes M, N, and K sizes defined in multiples of 8.  If they are not defined by multiples of 8, currently, codegen returns a tensor of zeroes.
- FP16 inputs are required but the output tensor cannot be casted back to FP16.  When attempting a cast, an error occurs in segmentation.
- The input tensors defined in Pytorch need to be contiguous in the intended layout.  Using the inputs as defined for Pytorch does not work.

For example, in order to define an NN layout example you need to do the following:
`M = 24, N = 16, K = 8`
For nvFuser:
```
nvf_inputs_nn = [
    torch.randn(8, 24, device="cuda", dtype=torch.float16),
    torch.randn(16, 8, device="cuda", dtype=torch.float16),
]
```
Eager.  Pytorch expects `input1 = [M, K] and input2 = [K, N]` which is why the transposes are necessary.
```
eager_inputs_nn = [
    torch.randn(8, 24, device="cuda", dtype=torch.float16).transpose(0, 1),
    torch.randn(16, 8, device="cuda", dtype=torch.float16).transpose(0, 1),
]
```

### CODE REVIEW

**Comment by kevinstephano** [comment]
!build

**Comment by kevinstephano** [comment]
I have a Volta error!  I am guessing I didn't handle something entirely correctly for Volta. 👀

```
Traceback (most recent call last):
  File "/opt/pytorch/nvfuser/python_tests/test_python_frontend.py", line 2122, in test_matmuls
    nvf_out, _ = self.exec_nvfuser(
  File "/opt/pytorch/nvfuser/python_tests/test_python_frontend.py", line 62, in inner
    test_fn(*args, **kwargs)
  File "/opt/pytorch/nvfuser/python_tests/test_python_frontend.py", line 98, in exec_nvfuser
    out = fd.execute(inputs)
  File "/opt/pytorch/pytorch/nvfuser/__init__.py", line 76, in execute
    result = self._execute(inputs, override_user_schedule)
RuntimeError: h.has_value() INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/fusion_segmenter.cpp":2993, please report a bug to PyTorch. Can not find a scheduler to schedule fusion segment
```

**Review by zasdfgbnm — APPROVED** [review]

**Comment by zasdfgbnm** [comment]
Is this because NN is not supported on Volta yet? I think you can just skip NN for now.

**Comment by kevinstephano** [comment]
> Is this because NN is not supported on Volta yet? I think you can just skip NN for now.

I tried to switch to using the TT case on Volta by switching the inputs and then adding a permute after the mma.  The issue looks to be the segmenter does not like the permute after the mma.

All the other layouts see this issue:

```
Traceback (most recent call last):
  File "/opt/pytorch/pytorch/nvfuser/__init__.py", line 76, in execute
    result = self._execute(inputs, override_user_schedule)
RuntimeError: status INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/scheduler/matmul_utils.cpp":531, please report a bug to PyTorch. Core part of heuristics failed to initialize.
```

**Comment by kevinstephano** [comment]
Hmm, it looks like the Heuristic just doesn't support Volta.  It would be nice if the assert said that explicitly.

```python
inline bool initCoreHeuristics(                                                                                                                                                                                                                                                            
    std::shared_ptr<MatmulParams> params,                                                                                                                                                                                                                                                  
    const MmaOptions::MacroType& mma_op,                                                                                                                                                                                                                                                   
    const ProblemShape& problem_shape) {                                                                                                                                                                                                                                                   
  const GemmTile instruction_tile = getMmaOpShape(mma_op);                                                                                                                                                                                                                                 
  GemmTile warp_tile = {-1, -1, -1};                                                                                                                                                                                                                                                       
  GemmTile cta_tile = {-1, -1, -1};                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                           
  using DimType = decltype(GemmTile::m);                                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                           
  // warp tile shape                                                                                                                                                                                                                                                                       
  {                                                                                                                                                                                                                                                                                        
    if (isAmpere(mma_op)) {                                                                                                                                                                                                                                                                
      // Initial target:                                                                                                                                                                                                                                                                   
      // - 1 MMA ops per thread in a warp (32 threads), warp tile should be                                                                                                                                                                                                                
      //   then 32x bigger than instruction tile,                                                                                                                                                                                                                                          
      // - start with [4, 4, 2] shape, later it should depend on problem                                                                                                                                                                                                                   
      //   shape and have bigger impact on CTA tile shape                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                           
      const DimType m_ratio = 4;                                                                                                                                                                                                                                                           
      const DimType n_ratio = 4;                                                                                                                                                                                                                                                           
      const DimType k_ratio = 2;                                                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                                           
      warp_tile = {                                                                                                                                                                                                                                                                        
          instruction_tile.m * m_ratio,                                                                                                                                                                                                                                                    
          instruction_tile.n * n_ratio,                                                                                                                                                                                                                                                    
          instruction_tile.k * k_ratio};                                                                                                                                                                                                                                                   
    } else {                                                                                                                                                                                                                                                                               
      // No support for Volta and Turing                                                                                                                                                                                                                                                   
----> return false;                                                                                                                                                                                                                                                                        
    }                                                                                                                                                                                                                                                                                      
  }       
 ```

**Comment by zasdfgbnm** [comment]
Haha, then probably just skip tests entirely for Volta.

**Comment by kevinstephano** [comment]
I narrowly enabled the `_matmul_xx` functions for just Ampere devices and altered the tests to appropriately reflect the asserts.

**Comment by kevinstephano** [comment]
!build

**Review by mmigdal-nv — APPROVED** [review]

**Comment by mmigdal-nv** [comment]
!build

**Comment by kevinstephano** [comment]
It looks like a general deserialization error is happening on V100.

```
Traceback (most recent call last):
  File "/opt/pytorch/nvfuser/python_tests/test_python_frontend.py", line 1819, in test_nanogpt_mha_dpa
    nvf_out, _ = self.exec_nvfuser(partial(nvfuser_fusion, prob=0.0), inputs)
  File "/opt/pytorch/nvfuser/python_tests/test_python_frontend.py", line 72, in inner
    fc.deserialize("foo.bin")
RuntimeError: Only the Ampere MMA Op is currently supported!

### [jit_python_tests_V100] TestNvFuserFrontend.test_nanogpt_split_mha_linears

Traceback (most recent call last):
  File "/opt/pytorch/nvfuser/python_tests/test_python_frontend.py", line [19](https://github.com/NVIDIA/Fuser/actions/runs/4873246468/jobs/8692548701#step:2:20)07, in test_nanogpt_split_mha_linears
    nvf_out, _ = self.exec_nvfuser(nvf_func, inputs)
  File "/opt/pytorch/nvfuser/python_tests/test_python_frontend.py", line 72, in inner
    fc.deserialize("foo.bin")
RuntimeError: Invalid scheduler query for id:[24](https://github.com/NVIDIA/Fuser/actions/runs/4873246468/jobs/8692548701#step:2:25)
```

File: python_tests/test_python_frontend.py [hunk 1]
```diff
@@ -2070,6 +2070,74 @@ def schedule(self):
...
+                self.assertEqual(eager_out, fp16_nvf_out)
+            else:
+                with self.assertRaisesRegex(
+                    RuntimeError, "Only the Ampere MMA Op is currently supported!"
+                ):
+                    nvf_out, _ = self.exec_nvfuser(
+                        partial(fusion_func, inps=nvf_test_inputs, matmul_fn=mm_str),
+                        nvf_test_inputs,
+                    )
+
```
Comment (rdspring1) [inline python_tests/test_python_frontend.py:2140]: Add `FusionCache.reset()` here after this test.

On Volta, the mma kernel is stored in the fusion cache, so subsequent tests will recreate the fusion that raises the exceptions. However, those tests won't have the `assertRaisesRegex` to catch the exception.

**Comment by kevinstephano** [comment]
!build

**Comment by kevinstephano** [comment]
!build

**Comment by kevinstephano** [comment]
!build

**Pull request merged**